### ISSUE TITLE
An unexpected verification result.

### ISSUE THREAD

**Issue by muchang** [seahorn/seahorn#155]
For this program:
```
void __VERIFIER_error(){ERROR:goto ERROR;}

int f(long a){return (--a > 0);}
int main(){if(f(0x80000000L)==0)__VERIFIER_error();return 0;}
```
The __VERIFIER_error() is infeasible, which could be comfirmed by clang or gcc.
But Seahorn gives `sat` verification result which means __VERIFIER_error() is feasible.

My command line is `~/seahorn/build/run/bin/sea pf <file_name>`
OS: `Ubuntu 16.04`
I am using the latest master branch version of Seahorn on GitHub.

Should I add some configurations to get the expected result?

**Comment by agurfinkel** [comment]
There are several issues that prevent SeaHorn from working as you expect on this example.

1. Signed integer overflow is undefined in C, so what is the expected result is not clear
2. SeaHorn in default mode uses arbitrary precision integers. Hence, it will not work as 
    expected when overflow is possible. Reasoning with machine integers is supported by 
    the bmc engine. Use `sea bfp --inline --bmc FILE.c` for that.
3. This particular example is discharged by the pre-processor. `InstCombine` simplifies
    `--a > 0` into `a > 1`. This is correct because overflow is undefined. After that, which engine
     SeaHorn is using does not matter.

In summary, SeaHorn happens to work correctly (but not as expected) on this example. 
In general, SeaHorn is not the best tool for programs with unsigned integer overflow.
It should be handled correctly by BMC engine, but not by the HORN engine.

**Comment by muchang** [comment]
Thanks for your response. In the case below, there is no overflow problem, but Seahorn still gives the unexpected result:
```
void __VERIFIER_error(){ERROR:goto ERROR;}

volatile int i = 24;
volatile int j = 30;
volatile int k = 1;

int main()
{
        float pri2 = (((float) i / j) * (10000 / 1000) * k);
        if (pri2 != 8.0) __VERIFIER_error();
	return 0;
}
```
The __VERIFIER_error() is also infeasible.
But Seahorn gives `sat` verification result which means __VERIFIER_error() is feasible.

I used both `~/seahorn/build/run/bin/sea bpf --bmc <file_name>` and `~/seahorn/build/run/bin/sea pf <file_name>` commandline on Ubuntu 16.04 and the latest verison of Seahorn.

**Comment by agurfinkel** [comment]
The current release does not support floats at all, even in BMC mode. There is an experimental branch with float support, but it has not been integrated into the main branch yet.

As far as I know, reasoning about floats is still very challenging for existing SMT-solvers. What kind of verification problems do you have that require reasoning about floats?

**Comment by muchang** [comment]
Thanks for your reply. I am just checking some features of Seahorn to ensure that Seahorn can work in some harsh situations. I found another case for which Seahorn gives an unexpected result:
```
void __VERIFIER_error(){ERROR:goto ERROR;}

int main ()
{
  char *s = ":a";
  char *lim = s + 2;
  char *ptr = s;
  int sword = 1;
  while (ptr < lim && sword--)
	{
	  while (*ptr != ':')
	    ++ptr;
    ptr++;
	}
  if (ptr != s + 1)
    __VERIFIER_error();
  return 0;
}
```
The __VERIFIER_error() is also infeasible.
But Seahorn gives `sat` verification result which means __VERIFIER_error() is feasible
The commandline is `~/seahorn/build/run/bin/sea pf <file_name>` on Ubuntu 16.04 and the latest verison of Seahorn.

But Seahorn works well with commandline `~/seahorn/build/run/bin/sea bpf <file_name>`.
